### DISCUSSION TITLE
Multiapp mesh transfer with adaptivity

### DISCUSSION THREAD

**Discussion by xueyang94** [idaholab/moose#22417] [Q&A General] [answered]
I have one mainapp that transfers a few variables from one subapp using FullSolveMultiApp. The transfer only occurs after the subapp is finished. In the subapp, I use mesh Adaptivity starting at initial step. The subapp domain is smaller than the mainapp domain.  How to transfer the subapp mesh after adaptivity? Thanks.

**Comment by GiudGiud** [comment, marked as answer]
Hello

we don’t transfer meshes. However if you want to refine the main app the same way as the subapp you can transfer the indicator and use it for the criterion for adaptivity

Guillaume

**Reply by GiudGiud** [reply]
the trickiness might be to order the transfer and the adaptivity correctly. 

Maybe if you run the mulitapp and do the transfer on timestep_end, then run the adaptivity on timestep_begin that would work for you?

**Reply by xueyang94** [reply]
Hello Guillaume, you have described what I am doing currently. I transfer the indicator (bnds) from the subapp to mainapp after the subapp is finished, and run the adaptivity of mainapp on timestep_begin. However, since the mainapp does not start with the finest mesh of the subapp, as the mesh refines, the mainapp would crash due to memory.

I want the mainapp to start with the finest mesh so that at least the crash would happen at the first timestep. The mainapp adaptivity block uses ValueRangeMarker on both the transferred indicator (bnds) and the variables in the mainapp. However, at the initial timestep of mainapp, only its own variables are refined, while bnds is at its coarsest. This is strange to me why bnds is not refined just like other variables.

**Reply by xueyang94** [reply]
The issue can be visualized using the input files I supplied in Discussion #22342. The meshes of subapp and mainapp at timestep zero are below. There are three phases in the mainapp from left to right. The subapp simulates the left-most phase but with two grains on top of each other. Although mesh cannot be transferred as you mentioned, I would expect the left phase in the mainapp to have a refined horizontal grain boundary because the bnds indicator from subapp is refined together with other variables in the mainapp at timestep zero.

Subapp at timestep zero:
![subapp](https://user-images.githubusercontent.com/31705995/198676074-f58993ec-8cf4-4ee6-9bd9-df953da8dfcf.PNG)

Mainapp at timestep zero:
![mainapp](https://user-images.githubusercontent.com/31705995/198676106-31dec7e8-5b78-479f-abd5-e77814ab9119.PNG)

**Reply by GiudGiud** [reply]
I ran your inputs and I think I have a better understanding of what you want but I think there are a few issues:

- in the main app I'm seeing etaa0 and etaa1 used as markers, not bnds. 
- the mesh of the main app will never see the updated value from the transfer of bnds from the main app because when the transfer is performed, it does not have nodes in the center/middle of the square.
Once the mesh is refined, it could, but at INITIAL when the transfer is performed it does not. 
I made `ny` be even, so that there's nodes on the horizontal middle of the left box, and got the expected bnds value in the center. 

but even then, looking at the FEProblemBase.C order of execution of things in initialSetup():
- markers are on line 750
- adaptivity is on line 850
- multiapps and transfers are on line 1055

so adaptivity is ran before the multiapps are, the values transfered and so on. So the marker would not be used on INITIAL. 

So your options:
- make a PR to MOOSE moving adaptivity after the first round of multiapps. I doubt anyone relies on the opposite
- run a fake time step with no physics at the beginning of your main app
- have a very very small first time step with a coarsened mesh. In a very small time step in the real world, nothing happens, and same in MOOSE.

**Reply by xueyang94** [reply]
Thanks for the reply. I used an updated input file where bnds was transferred instead of grain variables. The transfer does not have nodes in the center of the square, and it only have nodes on the edge of the square? Not sure how this transfer works. 

If the adaptivity is ran before the multiapps are, I would imagine the markers being used on INITIAL because the marker is associated with adaptivity.

**Reply by GiudGiud** [reply]
ok for updated input
The meshfunction or shapeevaluation transfer evaluates the source variable at the locations of the target mesh.
You do not have any mesh node at the right location in the target mesh at initialization

yes, markers are also pre-multiapp
